prereise/gather/flexibilitydata/bus/bus_data.py

```diff
@@ -0,0 +1,339 @@
+import os
+import pickle as pkl
+import time
+from math import acos, cos, degrees, radians, sin
+
+import geopy
+import numpy as np
+import pandas as pd
+import requests
+import scipy.io as spio
+from geopy.extra.rate_limiter import RateLimiter
+
+
+def get_bus_pos(case_path, cache_path):
+    """Read a .mat case and extract the lat/lon coordinate of all buses
+
+    :param str case_path: path to a .mat case file representing a power network
+    :param str cache_path: folder to store processed cache files
+    :return: (*numpy.ndarray*) -- a list of (busID, lat, lon) stored as a numpy array
+    """
+
+    case = spio.loadmat(case_path)
+    mpc = case["mpc"]
+
+    bus = mpc["bus"][0][0]
+    bus_num = bus.shape[0]
+    bus2sub = mpc["bus2sub"][0][0]
+    sub = mpc["sub"][0][0]
+    subid = mpc["subid"][0][0]
+
+    bus_pos = np.zeros((bus_num, 3))
+
+    # [id, lat, long]
+    subid_list = subid[:, 0].tolist()
+    for i in range(bus_num):
+        bus_sub_idx = subid_list.index(bus2sub[i, 0])
+        bus_pos[i, 0] = bus[i, 0]
+        bus_pos[i, 1] = sub[bus_sub_idx, 2]
+        bus_pos[i, 2] = sub[bus_sub_idx, 3]
+
+    with open(os.path.join(cache_path, "bus_pos.pkl"), "wb") as fh:
+        pkl.dump(bus_pos, fh)
+
+
+def get_bus_fips(bus_pos, cache_path, start_idx=0):
+    """Try to get FIPS of each bus in a case mat using FCC AREA API
+    Can take hours to run, save to cache file for future use
+
+    :param numpy.ndarray bus_pos: a list of (busID, lat, lon) stored as a numpy array
```

Comment:
> Following my previous comment above, if we expect a pandas data frame for `bus_pos` here, some of the operations below can be simplified and more readable, such as querying column via names rather than index, the `fips` information can be added via another column named `fips` rather than creating a dictionary, etc.
> 
> Same comments apply for `get_bus_zip` function below.

Although I have changed the input type to pd.dataframe, it does not seem practical to change the format of output to "add via another column" as you suggested. Many functions depend on the existing format of bus fips and bus zip cache in np.array, if I were to change the way they are stored as a df that contains bus coordinate and the additional columns I would need to refactor most functions I've got.

Reply:
The "readable indexing" is fixed but I will just leave the output as it is.